src/RFC-0321_ProcessingForeignProposals.md

```diff
@@ -0,0 +1,113 @@
+
+In a local shard committee, the proposed block **must** include a counter per epoch and per shard. If the proposal includes transactions that involve other shards, this counter **must** be incremented.
+
+When the proposed block becomes committed locally, the block **must** be broadcast to each involved shard that was incremented, along with evidence of being committeed (The chain of QCs must be included).
+
+To ensure this, each node in the local committee will forward this committed block, along with a 3 chain of QC's proving it was committeed.
+
+As a local committee member, when I receive a foreign proposal, if it is valid I will queue up a special command ForeignProposal(number, QC_Hash) that I must propose 
+when I am next leader (if it has not been proposed already). I also **should** request all transaction hashes that I have not seen from involved_shards for each transaction in the proposal, and add them to my mempool for execution.
+
+In the same order of transactions in the proposal, the transactions must be sequenced into the local chain as either a TIMEOUT or a LOCALPREPARE(TxId, ForeignShardId).
+
+When a LOCALPREPARE command is found in a local proposal, there **must** be a 
+Prepare(tx) command for that transaction in or before the current proposal.
+
+In the [TIMEOUT_TIME] block  (e.g. 1000 blocks, but should probably be equal to committee_size * X rounds) after the ForeignProposal command has been proposed, 
+all transactions in the proposal must be sequenced, either as LOCALPREPAREs or TIMEOUTs. This is to cater for the case where a transaction was prepared by a foreign committee, but was never found on the network. This is unlikely to happen 
+in a non-malicious scenario.
+
+For this reason, LOCAL_PREPARES and TIMEOUTs do not count to the block size.
+
+NOTE: ForeignProposal commands can be proposed in between a previous ForeignProposal and LocalPrepare/Timeout commands, but commands from the ForeignProposal **must** only be proposed after all transactions in the first ForeignProposal have been sequenced. 
+The TIMEOUT_TIME block is counted from the height where the ForeignProposal is sequenced.
+
+ForeignProposal commands **must** appear in strict ascending order in the blockchain, but do not have to be in sequential blocks. In other words, for shard *s*, the block containing ForeignProposal(*s*, 1) must have a height lower than ForeignProposal(*s*, 2). Also, if a chain contains ForeignProposal(*s*, 1) and ForeignProposal(*s*, 3), then it **must** also contain ForeignProposal(*s*, 2).
+
+If a node receives a foreign proposal (not the command), and it has not received
+the previous foreign proposal, then it should ask the committee to provide it to them.
+
+If the LOCAL_PREPARE command appears in a local proposal for a transaction and is the last such LOCAL_PREPARE and all shards have voted to accept, an ALL_PREPARED(tx) command **must** also be present in the local proposal.
```

Comment:
So if we have LOCAL_PREPARE(tx) and we have all foreign LOCAL_PREPARE already, we should put ALL_PREPARED(tx) instead? Or do you mean both will be present? What will happen when the block is too big and it's not enough space to add both?

Reply:
Maybe we can just put the ALL_PREPARED in then, and same for SOME_PREPARED and ALREADY_ABORTED